### ISSUE TITLE
Per-project CSS file inclusion

### ISSUE THREAD

**Issue by rbeezer** [rbeezer/mathbook#65]
Develop a way for an author to specify a CSS file that extends the stock style provided for HTML.  This would likey be a `param` as part of processing, since adding it to `<docinfo>` does not seem entirely right (the default files are given as param's).

**Comment by davidfarmer** [comment]
If the author specifying

"put theorems in a pink box"

"put a hollow square at the end of examples"

then perhaps those do belong in docinfo.  The result is custom CSS,
LaTeX, ePub, and whatever else.  That is what we eventualy want, since
it is painful for the author to develop customizations for each
possible medium.

**Comment by rbeezer** [comment]
I think I agree.  I certainly agree that *reasonable* options like this belong 
someplace and save the author the trouble of implementing lots of different 
versions of the same effect.

I think of a box, and its pinkness, as presentation, though it is universal 
across formats and likely a popular request.  So we should have an easy way for 
an author to ask for this presentation and we should then react to it across all 
formats.  But it is not *content*, so should not go into docinfo.

That said - no matter what we provide, I think there will always be a desire to 
do more and go further, especially if it is unique to a single project. 
Providing their own CSS would be the right way to make this happen for HTML.

So I see a general use for boxes, and for making them pink.  But if an author 
insists on a double rule before each theorem and a triple rule following, then 
maybe they should do that themselves.  There is a line there somewhere.

**Comment by dbrianwalton** [comment]
So docinfo right now includes LaTeX header information, packages that are needed, etc. So it seems to make sense to me that docinfo could include a URL to a CSS file. If the URL is a file, then it is copied to the working directory where the HTML files are dropped and is linked to each HTML file generated. If the URL is on a server, then that is linked. It would just be important that there were a clearly defined class assigned to each type of structured element in the document so that CSS can be used effectively.

For LaTeX, what if each of the elements had a custom environment that by default used whatever is standard. But using a formatting file (also identified in docinfo), the environments could be redefined however someone wanted, whether that means putting a box around the outside or putting a square directly below the last content.

**Comment by rbeezer** [comment]
Dear Brian,

"docinfo" is part of the source and so should be distinct from items used to 
affect style.  It may be giving the wrong impression right now - those latex 
items are required for source code used to generate images, so technically they 
are part of the document/source.  "docinfo" should have things that are repeated 
in various places, or that don't have a natural home in the normal hierarchy of 
the document.  So it is a bit of a dumping ground for oddball stuff (like the 
image you see in the upper-left of HTML pages that replaces the default 
book/article image).

I have a change almost ready to go that will enable the CSS addition that will 
do what you want - but the critical thing to realize is that you won't touch 
your source to make it happen.

Your suggestion about custom environments is exactly right.  We should 
definitely make an effort to do more things like that.  I'm OK with some latex 
code not being super-optimal originally, so long as the XML is being designed 
right.  Once the XML settles down then the latex can change without an author 
being aware, or needing to change their source.

Thanks for the input.

Rob

**Comment by rbeezer** [comment]
I've added a processing parameter that allows you to list as many CSS files as you like, local or on another server as part of processing.  For a big project, you'll want a thin XSL layer and specify this there.  For one-off and testing include the following on your command line

```--stringparam html.css.extra "one.css   http://server.com/two.css   three.css"```

Only allowed separator is space(s) (so no spaces in filenames) and you will need quotes to protect the spaces, and the slashes in external URLS.

See cdb3ecc2 for extra insight.



Undocumented and I have only tested that the URLs migrate to the HTML files as expected.
